notes/lossless-ast-parsing.md

```diff
@@ -0,0 +1,137 @@
+# Notes on a full fidelity parser
+Nickel does not have a formatter, which it should ideally have.
+The current parser and ast are not suited for this purpose since they perform
+desugaring during decoding and drop all information on whitespace and
+comments. For example, the following program:
+```nickel
+{
+  // This bar is very important
+  foo.bar = "bar",
+  foo.baz = "baz",
+}
+```
+
+Parses to:
+```nickel
+{
+  foo = { bar = "bar"} & { baz = "baz" },
+}
+```
+
+This note explores the options we have within the rust ecosystem to construct a
+full fidelity (or lossless) AST.
+
+# Keeping the existing lalrpop grammar
+The current lexing/parsing stack of Nickel is based on a lexer generated using
+[logos][logos], a parser generated by
+[lalrpop][lalrpop], and a native AST.
+
+Our initial idea was to keep the existing lalrpop grammar in place, but
+replacing the AST with a [rowan][rowan] based
+alternative. Unfortunately, while it is possible to integrate logos, rowan, and
+lalrpop, comments and whitespace would still be lost. This is an inherent
+problem with lalrpop in which grammar rules match directly on the token input.
+Hence, dealing with whitespace for an operator 4 variants (whitespace optional
+on both sides).
+
+This thus implies that lalrpop has to be replaced.
+
+# Options
+## Custom
+Like rustc, nickel could employ a custom lexer and parser.
+
+### Advantages
+- Full control
+
+### Disadvantages
+- By far the most work
+- Rapid development is harder when implementing a parser from scratch
+
+## [Rowan][rowan] + [Ungrammar][ungrammar] (optional) + [Logos][logos]
+Rowan and Ungrammar are two libraries created for the
+[rust-analyzer][rust-analyzer] project.
+
+Of course, rust-analyzer is greatly benefitted by a lossless ast (CSTs as
+they call them). To this extend they created two libraries specifically for
+this purpose. [rowan][rowan] and [ungrammar][ungrammar].
+
+Rowan is a widespread library that has already seen use in existing languages.
+For instance for the
+[slint](https://github.com/slint-ui/slint/) language that make use of rowan for
+their language server and
+[compiler](https://github.com/slint-ui/slint/tree/master/internal/compiler).
+Hence we can safely say rowan is likely suitable for nickel.
+
+Ungrammar is a grammar that works to construct the SyntaxNodes as described in
+[syntax overview](https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/syntax.md)
+of rust-analyzer. Ungrammar has not seen use outside of rust-analyzer. I
+personally do not see a need for Nickel to base its SyntaxNodes on ungrammar.
+
+Since we can no longer use lalrpop, the parser would have to be written by
+hand. Parser combinators are not very suitable to this approach, as many
+are based around a parser succeeding or failing, negating rowan's advantage of
+being able to produce an AST even when there is an error.
+
+There is of course nothing fundamentally impossible about using some parser
+combinator library.
```

Comment:
Another argument used usually against parser combinators is that they are not very efficient, but I don't know if that claim is fair in practice. I imagine that it depends a lot on the particular parser combinator library.